src/crimson/os/seastore/transaction.h

```diff
@@ -76,6 +76,18 @@ class Transaction {
     ceph_assert(inserted);
   }
 
+  //XXX: boost::intrusive::set seems to have a bug in it, in which
+  //	 it may leak the first item when trying to find a specific item
+  CachedExtent::index::iterator find_in_write_set(paddr_t addr) {
+    auto it = write_set.begin();
+    for (; it != write_set.end(); ++it) {
```

Comment:
i'd avoid iterating through a binary search tree for looking up with a given key. since the underlying data structure is a red-black tree, we should be able to find the value with an O(log(n)) operation. would you please create a unit test for `boost::intrusive::set` and
```c++
template<class KeyType, class KeyTypeKeyCompare>
   iterator find(const KeyType& key, KeyTypeKeyCompare comp);
```

Reply:
Yeah, I'd like more information on the bug in question.  The reason to use boost intrusive set here is specifically for the log n lookup, and we use it elsewhere as well.

Reply:
@tchaikov @athanatos I haven't been able to reproduce the case with a simple unit test, but I can do it when doing the "unittest-transaction-manager" unit test:

First, I added more output to Transaction and Cache code:
```
  get_extent_ret get_extent(paddr_t addr, CachedExtentRef *out) {
    ::crimson::get_logger(ceph_subsys_seastore).debug("{} {}", __func__, addr);
    if (retired_set.count(addr)) {
      return get_extent_ret::RETIRED;
    } else if (auto iter = write_set.find_offset(addr);
        iter != write_set.end()) {
      ::crimson::get_logger(ceph_subsys_seastore).debug("{} found in write_set {}", __func__, *iter);
      if (out)
        *out = CachedExtentRef(&*iter);
      return get_extent_ret::PRESENT;
    } else if (
      auto iter = read_set.find(addr);
      iter != read_set.end()) {
      ::crimson::get_logger(ceph_subsys_seastore).debug("{} found in read_set {}", __func__, *iter->ref);
      if (out)
        *out = iter->ref;
      return get_extent_ret::PRESENT;
    } else {
      return get_extent_ret::ABSENT;
    }
  }
```

```
  using get_extent_iertr = base_iertr;
  template <typename T>
  get_extent_iertr::future<TCachedExtentRef<T>> get_extent(
    Transaction &t,
    paddr_t offset,
    segment_off_t length) {
    CachedExtentRef ret;
    auto result = t.get_extent(offset, &ret);
    if (result != Transaction::get_extent_ret::ABSENT) {
      assert(result != Transaction::get_extent_ret::RETIRED);
      ::crimson::get_logger(ceph_subsys_seastore).debug("{} {} extent: {}", __func__, (void*)&t, *ret);
      t.dump_write_set(offset);
      return seastar::make_ready_future<TCachedExtentRef<T>>(
        ret->cast<T>());
    } else { 
      return trans_intr::make_interruptible(
        get_extent<T>(offset, length)
      ).si_then(
        [&t](auto ref) mutable {
          if (!ref->is_valid()) {
            t.conflicted = true;
            return get_extent_iertr::make_ready_future<TCachedExtentRef<T>>();
          } else {
            t.add_to_read_set(ref);
            return get_extent_iertr::make_ready_future<TCachedExtentRef<T>>(
              std::move(ref));
          }
        });
    } 
  }
```

```
  void dump_write_set(paddr_t addr) {
    for (auto it = write_set.begin();
         it != write_set.end();
         ++it) {
        if (it->poffset == addr) {
          ::crimson::get_logger(ceph_subsys_seastore).debug(
              "{} write_set: {}", (void*)this, *it);
        }
    }
  } 
```

And in the output of unittest-transaction-manager, I found the following lines:

```
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - BtreeLBAManager::get_root: reading root at paddr_t<11, 536576> depth 2
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - get_lba_btree_extent: reading internal at offset paddr_t<11, 536576>, depth 2
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - get_extent paddr_t<11, 536576>
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - get_extent found in read_set CachedExtent(addr=0x2ae3e865a420, type=LADDR_INTERNAL, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<11, 536576>, state=CLEAN, last_committed_crc=2216781921, refcount=3, size=12, meta=btree_node_meta_t(begin=0, end=18446744073709551615, depth=2))
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - get_extent 0x2ae3e85bab70 extent: CachedExtent(addr=0x2ae3e865a420, type=LADDR_INTERNAL, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<11, 536576>, state=CLEAN, last_committed_crc=2216781921, refcount=4, size=12, meta=btree_node_meta_t(begin=0, end=18446744073709551615, depth=2))
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - 0x2ae3e85bab70 write_set: CachedExtent(addr=0x2ae3e868aab0, type=LADDR_INTERNAL, version=1, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<11, 536576>, state=MUTATION_PENDING, last_committed_crc=2216781921, refcount=1, size=12, meta=btree_node_meta_t(begin=0, end=18446744073709551615, depth=2))
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - mutate_mapping_internal, CachedExtent(addr=0x2ae3e865a420, type=LADDR_INTERNAL, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<11, 536576>, state=CLEAN, last_committed_crc=2216781921, refcount=4, size=12, meta=btree_node_meta_t(begin=0, end=18446744073709551615, depth=2)) 1949696 -> paddr_t<8, 28672>
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - get_lba_btree_extent: reading leaf at offset paddr_t<8, 28672>, depth 1
DEBUG 2021-07-04 16:45:17,807 [shard 0] seastore - get_extent paddr_t<8, 28672>
unittest-transaction-manager: ../src/crimson/os/seastore/cache.h:276: crimson::interruptible::interruptible_errorator<crimson::os::seastore::TransactionConflictCondition, crimson::errorator<crimson::unthrowable_wrapper<const std::error_code&, ((const std::error_code&)(& crimson::ec<std::errc::io_error>))> > >::future<boost::intrusive_ptr<T> > crimson::os::seastore::Cache::get_extent(crimson::os::seastore::Transaction&, crimson::os::seastore::paddr_t, crimson::os::seastore::segment_off_t) [with T = crimson::os::seastore::lba_manager::btree::LBALeafNode; crimson::interruptible::interruptible_errorator<crimson::os::seastore::TransactionConflictCondition, crimson::errorator<crimson::unthrowable_wrapper<const std::error_code&, ((const std::error_code&)(& crimson::ec<std::errc::io_error>))> > >::future<boost::intrusive_ptr<T> > = crimson::interruptible::interruptible_future_detail<crimson::os::seastore::TransactionConflictCondition, crimson::errorator<crimson::unthrowable_wrapper<const std::error_code&, ((const std::error_code&)(& crimson::ec<std::errc::io_error>))> >::_future<crimson::errorated_future_marker<boost::intrusive_ptr<crimson::os::seastore::lba_manager::btree::LBALeafNode> > > >; crimson::os::seastore::segment_off_t = long int]: Assertion `result != Transaction::get_extent_ret::RETIRED' failed.
```

`CachedExtent(addr=0x2ae3e865a420)` is the old extent, while `CachedExtent(addr=0x2ae3e868aab0)` is the new one. According to the output, `Transaction::get_extent()` failed to find the new one in write_set, which led to the old one being returned, while the new one is actually in the write_set. This is how I came to the idea that intrusive set might leak some item in the find method.

Reply:
@xxhdx1985126 i revised your change by adding an assert() statement so the test fails if it finds an extent in read_set but an extent with the same paddr also exists in the write_set. see https://github.com/tchaikov/ceph/tree/wip-crimson-set-find. i ran the test for 10 times without failures.

is this issue reproducible consistently?

Reply:
@tchaikov I can reproduce this issue in the extent placement manager PR, but I think this issue is independent from the extent placement related code.

Reply:
@xxhdx1985126 would you mind test my PR to see if it fails in your environment?

Reply:
sure:-)

Reply:
No, will test it

Reply:
@tchaikov I can't reproduce the issue...

Reply:
@xxhdx1985126 are you testing debug build or release build?

Reply:
debug build

Reply:
then i'd like to see a reproducer for understand this issue better.

Reply:
@tchaikov this branch should be able to steadily reproduce the issue: https://github.com/xxhdx1985126/ceph/tree/wip-extent-placement-manager-4

Reply:
@tchaikov As discussed offline, I modified the `Transaction::get_extent()` of the branch https://github.com/xxhdx1985126/ceph/tree/wip-extent-placement-manager-4 as follows:

```
  get_extent_ret get_extent(paddr_t addr, CachedExtentRef *out) {
    ::crimson::get_logger(ceph_subsys_seastore).debug("{} {}", __func__, addr);
    if (retired_set.count(addr)) {
      return get_extent_ret::RETIRED;
    } else if (auto iter = write_set.find_offset(addr);
        iter != write_set.end()) {
      ::crimson::get_logger(ceph_subsys_seastore).debug("{} found in write_set {}", __func__, *iter);
      if (out)
        *out = CachedExtentRef(&*iter);
      return get_extent_ret::PRESENT;
    } else if (auto iter = find_in_write_set(addr);
        iter != write_set.end()) {
      assert("shouldn't reach this point" == 0);
    } else if (
      auto iter = read_set.find(addr);
      iter != read_set.end()) {
      ::crimson::get_logger(ceph_subsys_seastore).debug("{} found in read_set {}", __func__, *iter->ref);
      if (out)
        *out = iter->ref;
      return get_extent_ret::PRESENT;
    } else {
      return get_extent_ret::ABSENT;
    }
  }
```

And the unittest-transaction-manager failed at the line "assert("shouldn't reach this point" == 0);":

```
DEBUG 2021-07-05 11:39:54,094 [shard 0] seastore - rewrite_logical_extent_p2 CachedExtent(addr=0x2b39dc5c9c50, type=TEST_BLOCK, version=1, dirty_from_or_retired_at=journal_seq_t(segment_seq=8, offset=paddr_t<10, 4096>), paddr=paddr_t<1, 3432448>, state=DIRTY, last_committed_crc=496849168, refcount=8, laddr=1683456, pin=LBAPin(1683456~4096->paddr_t<1, 3432448>) -> CachedExtent(addr=0x2b39dc6d0370, type=TEST_BLOCK, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<9, 278528>, state=INITIAL_WRITE_PENDING, last_committed_crc=0, refcount=2, laddr=1683456, pin=LBAPin(1683456~4096->paddr_t<1, 3432448>)
DEBUG 2021-07-05 11:39:54,094 [shard 0] seastore - Cache::get_root(0x2b39dc5a5790): root already on transaction CachedExtent(addr=0x2b39dc609000, type=ROOT, version=2, dirty_from_or_retired_at=journal_seq_t(segment_seq=9, offset=paddr_t<11, 532480>), paddr=paddr_t<NULL_SEG, NULL_OFF>, state=DIRTY, last_committed_crc=0, refcount=5)
DEBUG 2021-07-05 11:39:54,094 [shard 0] seastore - BtreeLBAManager::get_root: reading root at paddr_t<11, 536576> depth 2
DEBUG 2021-07-05 11:39:54,094 [shard 0] seastore - get_lba_btree_extent: reading internal at offset paddr_t<11, 536576>, depth 2
DEBUG 2021-07-05 11:39:54,094 [shard 0] seastore - get_extent paddr_t<11, 536576>
unittest-transaction-manager: ../src/crimson/os/seastore/transaction.h:48: crimson::os::seastore::Transaction::get_extent_ret crimson::os::seastore::Transaction::get_extent(crimson::os::seastore::paddr_t, crimson::os::seastore::CachedExtentRef*): Assertion `"shouldn't reach this point" == 0' failed.
Aborting on shard 0.
```

Reply:
Ah, I see, this is not the bug of `boost::intrusive::set`, it is because we abruptly update the extent paddr in-place, causing incorrect order in `write_set`.

```
get_extent paddr_t<11, 536576>
seastore - 0x60109a7e3680 write_set: CachedExtent(addr=0x60109a7e5200, type=LADDR_INTERNAL, version=1, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<11, 536576>, state=MUTATION_PENDING, last_committed_crc=1636825732, refcount=1, size=12, meta=btree_node_meta_t(begin=0, end=18446744073709551615, depth=2))
seastore - 0x60109a7e3680 write_set: CachedExtent(addr=0x60109ac5a500, type=TEST_BLOCK, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<9, 282624>, state=INITIAL_WRITE_PENDING,  last_committed_crc=0, refcount=2, laddr=2506752, pin=LBAPin(2506752~4096->paddr_t<1, 5103616>)
seastore - 0x60109a7e3680 write_set: CachedExtent(addr=0x60109ac5a140, type=TEST_BLOCK, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<9, 286720>, state=INITIAL_WRITE_PENDING, last_committed_crc=0, refcount=1, laddr=311296, pin=LBAPin(311296~4096->paddr_t<1, 634880>)
seastore - 0x60109a7e3680 write_set: CachedExtent(addr=0x60109ac59b80, type=TEST_BLOCK, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<9, 290816>, state=INITIAL_WRITE_PENDING, last_committed_crc=0, refcount=1, laddr=2134016, pin=LBAPin(2134016~4096->paddr_t<1, 4349952>)
...
```
The order paddr_t<11, 536576>, paddr_t<9, 282624>, paddr_t<9, 286720>, paddr_t<9, 290816> ... in `write_set` is incorrect.

Reply:
Oh, yes, you are right.

Reply:
I'm closing this PR, as it is wrong.